### PR TITLE
Time Series HTTP Server

### PR DESCRIPTION
**Pull request by mrtracy**
Add HTTP request serving to the ts package. This currently only exposes a Query
method.

As part of this commit, the protobuffers for time series have been slightly
refactored to emphasize a "Request" and "Response" - the structure of these
messages has been chosen to resemble the standard Request and Response messages
used by other currently popular time series databases (OpenTSDB, KairosDB). The
request structure allows for multiple sub-queries in the same batch.

This commit also introduces the TestHTTPSession helper structure in testutils -
this is intended for usage by tests which execute several simple HTTP statements
against the same server, where no errors are expected in the HTTP layer.

### CODE REVIEW

File: ts/query.go [hunk 1]
```diff
@@ -369,12 +369,12 @@ func (is unionIterator) sumAvg() float64 {
...
-// will represent the sum across all sources at the same sample period. The
-// response will contain a list of all sources which were summed to produce the
-// result.
+// If data for the named time series was collected from multiple sources, each
+// returned datapoint will represent the sum of datapoints from all sources at
+// the same time. The returned string slices contains a list of all sources for
+// the metric which were aggregated to produce the result.
 func (db *DB) Query(name string, r Resolution, startNanos, endNanos int64) (
-	*proto.TimeSeriesQueryResult, error) {
+	[]*proto.TimeSeriesDatapoint, []string, error) {
```
Comment (mrtracy) [inline ts/query.go]: I believe I should push the Resolution type into the protos package, which would let me pass a proto.TimeSeriesQueryRequest to this method. However, i've decided to do that in another PR, because this one already has a bit of refactoring.

File: server/testserver.go [hunk 2]
```diff
@@ -99,6 +100,14 @@ func (ts *TestServer) Clock() *hlc.Clock {
 	return nil
 }
 
+//TsDB returns the ts.DB instance used by the TestServer.
+func (ts *TestServer) TsDB() *ts.DB {
+	if ts != nil {
```
Comment (tamird) [inline server/testserver.go]: how can the receiver be nil?

File: proto/timeseries.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+// Note that this method does not accumulate data into individual samples in the
+// case where multiple datapoints fall into the same sample period. Internal
+// data should be merged into the cockroach data store before being read; the
+// engine is responsible for accumulating samples.
+//
+// The returned slice of InternalTimeSeriesData objects will not be sorted.
+//
+// For more information on how time series data is stored, see
+// InternalTimeSeriesData and its related structures.
+func (ts TimeSeriesData) ToInternal(keyDuration int64, sampleDuration int64) (
```
Comment (tamird) [inline proto/timeseries.go]: there's a lot of relocated code here, and it's hard to distill what's worth reviewing from what's not. in the future, could you put relocation such as this in its own commit, with the rest of the changes following in their own commit?

File: proto/timeseries.proto [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    // for this query.
+    optional int64 start_nanos = 1 [(gogoproto.nullable) = false];
+    // A timestamp in nanoseconds which defines the late bound of the time span
+    // for this query. Must be greater than start_nanos.
+    optional int64 end_nanos = 2 [(gogoproto.nullable) = false];
+
+    // Each Query defines a specific metric to query over the time span of
+    // this request.
+    message Query {
+        // The exact name of the time series to query.
```
Comment (spencerkimball) [inline proto/timeseries.proto]: s/exact //

The word "exact" makes me wonder about non-exact specifications which don't exist (yet?).

File: proto/timeseries.proto [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        // A string which uniquely identifies the variable from which this data was
+        // measured.
+        optional string name = 1 [(gogoproto.nullable) = false];
+        // A list of sources from which the data was aggregated.
+        repeated string sources = 2 [(gogoproto.nullable) = false];
+        // Datapoints describing the queried data.
+        repeated TimeSeriesDatapoint datapoints = 3;
+    }
+
+    // A set of Results; there should be one result for each Query in the
```
Comment (spencerkimball) [inline proto/timeseries.proto]: Maybe firm up your comment a bit:

    // A set of Results; there will be exactly one result for each Query in the
    // matching TimeSeriesQueryRequest, in the same order.

Also, what happens in the case of a name which doesn't exist in the query? Just an empty result? You should document what happens here (and make sure you have a unittest for that case).

File: server/server.go [hunk 6]
```diff
@@ -193,6 +195,7 @@ func (s *Server) initHTTP() {
 	s.mux.Handle(kv.RESTPrefix, s.kvREST)
 	s.mux.Handle(kv.DBPrefix, s.kvDB)
 	s.mux.Handle(structured.StructuredKeyPrefix, s.structuredREST)
+	s.mux.Handle(ts.UrlPrefix, s.tsServer)
```
Comment (spencerkimball) [inline server/server.go]: Usually you'd make this URLPrefix in Go, no?

File: ts/server.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+			return
+		}
+		response.Results = append(response.Results, &proto.TimeSeriesQueryResponse_Result{
+			Name:       q.Name,
+			Sources:    sources,
+			Datapoints: datapoints,
+		})
+	}
+
+	// Marshal and return response
```
Comment (spencerkimball) [inline ts/server.go]: Punctuation, matt!

**Comment by spencerkimball** [comment]
LGTM

File: server/server.go (hunk 6 above)
Comment (mrtracy) [inline server/server.go]: Capitalized URL in these variables. I didn't rename the other prefixes in this change, but I do think we could use a consistency sweep through our server components.

File: ts/server.go (hunk 7 above)
Comment (mrtracy) [inline ts/server.go]: I am ashamed.

File: testutils/http.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+// byte slice.
+func (ths *TestHTTPSession) Post(relative string, body []byte) []byte {
+	req, err := http.NewRequest("POST", ths.baseURL+relative, bytes.NewBuffer(body))
+	if err != nil {
+		ths.t.Fatal(err)
+	}
+
+	// All requests currently accept JSON.
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
```
Comment (bdarnell) [inline testutils/http.go]: Everything from this line down looks like it's shared between Get and Post and could be factored into a separate method

**Comment by bdarnell** [comment]
LGTM
